docs/toctoc/guides/token-based-recovery-flow.md

```diff
@@ -0,0 +1,180 @@
+---
+id: token-based-recovery-flow
+title: Token-based Recovery Flow
+---
+
+This flow allows to implement the password recovery functionality of a web
+application.
+
+## Authentication domains
+
+The flow defines and orchestrates two
+[authentication domains](toctoc/concepts/functional-model.md):
+
+- **`LoginDomain`**: an authentication domain whose credential is the `Login`
+  data type.
+- **`AccessTokenDomain`**: an authentication domain whose credential is the
+  `AccessToken` data type.
+
+The `Login` data type is defined by a username and a password, while the
+`AccessToken` data type is defined by a value -- which is usually a randomly
+generated string -- and an expiration date.
+
+> 👉 Note: both the token-based authentication flow and the token-based recovery
+> flow use the same kind of domain (`AccessTokenDomain`).
+>
+> However, you should normally have **two separate instances** for those
+> domains, as their semantics and policies are likely to be different.
+>
+> For example, you usually want to limit the password recovery access token to
+> the recovery functionality, and use an auth token for anything else.
+
+## Functionalities
+
+The flow is implemented by the `TokenBasedRecoveryFlow` class, which exposes
+these functionalities:
+
+- `registerSubjectLogin`: registers a `Subject` for recovery, by creating an
+  `AccessToken` in the `AccessTokenDomain`. This is typically used when a user
+  begins the password recovery flow.
+
+- `recoverLogin`: resets the `Login` credentials for a `Subject`. The `Subject`
```

Comment:
This basically performs an update of the existing `Login` for a given `Subject`: internally, is this implemented through an atomic `update` or by unregistering and registering a new login? 
I'm asking because we already experienced a bug related to this: a user was deregistered on reset password but the subsequent registration failed and the user was locked out because the previous Login was already canceled (the reason why we opened https://github.com/buildo/retro/issues/12)

Reply:
It's currently not atomic. We can discuss this in more details in #12.

By "locked out" I assume you mean that they had to do a new password recovery, or were they prevented to do that as well?

Reply:
I remember that they were prevented to do that as well, once locked out. But we can discuss it more in details in the dedicated issue.

Reply:
I remember it was locked out and prevented to do that as weel, but we can discuss it better in #12